File: mobile/app/src/main/java/edu/cit/canadilla/wildcatslounge/mobile/core/util/OrderDisplay.kt
package edu.cit.canadilla.wildcatslounge.mobile.core.util

import edu.cit.canadilla.wildcatslounge.mobile.core.model.CartItemData
import edu.cit.canadilla.wildcatslounge.mobile.core.model.GroupedCartItem
import edu.cit.canadilla.wildcatslounge.mobile.core.model.GroupedOrderItem
import edu.cit.canadilla.wildcatslounge.mobile.core.model.MenuItemData
import edu.cit.canadilla.wildcatslounge.mobile.core.model.OrderItemData

object OrderDisplay {
	private fun normalizeName(value: String?): String = (value ?: "").trim().lowercase()

	fun createAddonNameSet(menuItems: List<MenuItemData>): Set<String> =
		menuItems.filter { MenuHelpers.isAddonItem(it) }
			.map { normalizeName(it.name) }
			.toSet()

	fun formatCustomizationMeta(
		servingType: String?,
		sugarLevelPercent: Int?,
		customizationNotes: String?
	): String? {
		val details = mutableListOf<String>()
		if (!servingType.isNullOrBlank() && servingType != "NONE") {
			details.add(servingType.lowercase().replaceFirstChar { it.uppercase() })
		}
		if (sugarLevelPercent != null) {
			details.add("$sugarLevelPercent% sugar")
		}
		val notes = customizationNotes?.trim()
		if (!notes.isNullOrBlank()) {
			details.add(notes)
		}
		return details.takeIf { it.isNotEmpty() }?.joinToString(" — ")
	}

	fun groupCartItems(items: List<CartItemData>, addonNames: Set<String>): List<GroupedCartItem> {
		val grouped = mutableListOf<GroupedCartItem>()
		val byId = mutableMapOf<Long, GroupedCartItem>()
		var currentMain: GroupedCartItem? = null

		for (item in items) {
			if (item.parentItemId != null) continue
			val isAddon = addonNames.contains(normalizeName(item.itemName))
			val entry = GroupedCartItem(item, mutableListOf())
			grouped.add(entry)
			byId[item.id] = entry
			if (!isAddon) currentMain = entry
		}

		for (item in items) {
			val parentId = item.parentItemId
			if (parentId != null) {
				byId[parentId]?.addons?.add(item)
				continue
			}
			if (byId.containsKey(item.id)) continue

			val isAddon = addonNames.contains(normalizeName(item.itemName))
			if (isAddon && currentMain != null) {
				currentMain.addons.add(item)
				continue
			}
			val entry = GroupedCartItem(item, mutableListOf())
			grouped.add(entry)
			byId[item.id] = entry
			if (!isAddon) currentMain = entry
		}

		return grouped
	}

	fun groupOrderItems(items: List<OrderItemData>, addonNames: Set<String>): List<GroupedOrderItem> {
		val grouped = mutableListOf<GroupedOrderItem>()
		val byId = mutableMapOf<Long, GroupedOrderItem>()
		var currentMain: GroupedOrderItem? = null

		for (item in items) {
			if (item.parentItemId != null) continue
			val isAddon = addonNames.contains(normalizeName(item.itemName))
			val entry = GroupedOrderItem(item, mutableListOf())
			grouped.add(entry)
			byId[item.id] = entry
			if (!isAddon) currentMain = entry
		}

		for (item in items) {
			val parentId = item.parentItemId
			if (parentId != null) {
				byId[parentId]?.addons?.add(item)
				continue
			}
			if (byId.containsKey(item.id)) continue

			val isAddon = addonNames.contains(normalizeName(item.itemName))
			if (isAddon && currentMain != null) {
				currentMain.addons.add(item)
				continue
			}
			val entry = GroupedOrderItem(item, mutableListOf())
			grouped.add(entry)
			byId[item.id] = entry
			if (!isAddon) currentMain = entry
		}

		return grouped
	}
}

File: web/lib/order-display.ts
import type { CartItem, MenuItem, OrderItem } from './api';
import { isAddonItem, normalizeMenuItems } from './menu';

type DisplayItem = {
  id: number;
  itemName: string;
  customizationNotes?: string;
  parentItemId?: number | null;
  servingType?: string | null;
  sugarLevelPercent?: number | null;
};

export type GroupedDisplayItem<T extends DisplayItem> = T & {
  addons: T[];
  standaloneAddon?: boolean;
};

function normalizeName(value: string | undefined): string {
  return (value || '').trim().toLowerCase();
}

export function createAddonNameSet(menuItems: MenuItem[]): Set<string> {
  return new Set(
    normalizeMenuItems(menuItems)
      .filter(isAddonItem)
      .map((item) => normalizeName(item.name))
  );
}

export function formatCustomizationMeta(item: DisplayItem): string | null {
  const details: string[] = [];
  if (item.servingType && item.servingType !== 'NONE') {
    details.push(item.servingType.charAt(0) + item.servingType.slice(1).toLowerCase());
  }
  if (typeof item.sugarLevelPercent === 'number') {
    details.push(`${item.sugarLevelPercent}% sugar`);
  }
  if (item.customizationNotes?.trim()) {
    details.push(item.customizationNotes.trim());
  }
  return details.length > 0 ? details.join(' — ') : null;
}

export function groupOrderItems(orderItems: OrderItem[], addonNames: Set<string>): GroupedDisplayItem<OrderItem>[] {
  return groupDisplayItems(orderItems, addonNames);
}

export function groupCartItems(cartItems: CartItem[], addonNames: Set<string>): GroupedDisplayItem<CartItem>[] {
  return groupDisplayItems(cartItems, addonNames);
}

function groupDisplayItems<T extends DisplayItem>(items: T[], addonNames: Set<string>): GroupedDisplayItem<T>[] {
  const grouped: GroupedDisplayItem<T>[] = [];
  const byId = new Map<number, GroupedDisplayItem<T>>();
  let currentMainItem: GroupedDisplayItem<T> | null = null;

  for (const item of items) {
    if (item.parentItemId != null) {
      continue;
    }

    const isAddon = addonNames.has(normalizeName(item.itemName));
    const nextItem: GroupedDisplayItem<T> = {
      ...item,
      addons: [],
      standaloneAddon: isAddon,
    };
    grouped.push(nextItem);
    byId.set(item.id, nextItem);
    currentMainItem = isAddon ? currentMainItem : nextItem;
  }

  for (const item of items) {
    if (item.parentItemId != null) {
      const parent = byId.get(item.parentItemId);
      if (parent) {
        parent.addons.push(item);
        continue;
      }
    }

    if (item.parentItemId == null && byId.has(item.id)) {
      continue;
    }

    const isAddon = addonNames.has(normalizeName(item.itemName));
    if (isAddon && currentMainItem) {
      currentMainItem.addons.push(item);
      continue;
    }

    const nextItem: GroupedDisplayItem<T> = {
      ...item,
      addons: [],
      standaloneAddon: isAddon,
    };
    grouped.push(nextItem);
    if (item.parentItemId == null) {
      byId.set(item.id, nextItem);
      currentMainItem = isAddon ? currentMainItem : nextItem;
    }
  }

  return grouped;
}
